misc: Fix EnrichedTuple rebuilding

enwask opened the pull request:
Fixes `EnrichedTuple` (and its hierarchy) rebuilding to an empty instance.

mloubout commented on devito/tools/data_structures.py:
```diff
@@ -43,10 +43,15 @@ class EnrichedTuple(tuple, Pickable):
     A tuple with an arbitrary number of additional attributes.
     """
 
+    __rargs__ = ('*values',)
```
keep consistent with signature so rename as items or change to values in __new__

FabioLuporini replied:
seconded, this should be called "items" though in practice it might work the same since it's prefixed with "*"

mloubout commented on devito/tools/data_structures.py:
```diff
@@ -43,10 +43,15 @@ class EnrichedTuple(tuple, Pickable):
         obj = super().__new__(cls, items)
         obj.__dict__.update(kwargs)
-        obj._getters = OrderedDict(zip(getters or [], items))
+        if getters is not None or '_getters' not in kwargs:
```
can `_getters

mloubout replied:
Also this will be problematic if `getters` is `None` because `obj._getters` will not exist, that's what the `getters or ()` is for so that `obj._getters` always exists.

FabioLuporini replied:
I don't see why adding an `if` would be necessary?

enwask replied:
My reasoning was that otherwise, if reconstructed by something other than `Reconstructable._rebuild` (I believe there's some pickling system that uses `__getnewargs_ex__` instead, no?) then `kwargs['_getters']` will have the old value to be unpickled, but `getters` will equal `None`. Then if we don't do this check, the constructor will overwrite whatever the old value obtained from unpicking was, and instead overwrite with an empty dict

enwask replied:
Wait that was my old reasoning, I forgot I changed the way I did this so that's not accurate anymore. I'll take a look again in a minute

enwask replied:
Well, have found at least one case of extra kwargs being used `left` and `right` passed to a `DimensionTuple` constructor from `AbstractFunction`. I feel like in the long run that really be specialized rather than abusing the object `__dict__` this way but I'll leave it for now. 

Which means that we do need this check still, I think

FabioLuporini commented on devito/tools/data_structures.py:
```diff
@@ -76,6 +81,16 @@ def __getnewargs_ex__(self):
         return self._getters.get(key, val)
 
+    @property
+    def values(self) -> tuple:
```
"items"

FabioLuporini commented on devito/types/utils.py:
```diff
@@ -24,6 +24,12 @@ class Stagger(Tag):
 
+    __rkwargs__ = EnrichedTuple.__rkwargs__ + ('left', 'right')
+
+    def __new__(cls, *args, getters=None, left=None, right=None, **kwargs):
```
if we really have to add this `__new__`, which basically doesn't add anything to the super()'s `__new__`, then it looks odd

enwask replied:
Doesn't need to be added to `__new__` so those parameters are explicit, but it's not strictly necessary since that's covered by `self.__dict__.assign(...` in the parent. I can remove the parameters here if preferred

mloubout commented on devito/tools/data_structures.py:
```diff
@@ -71,11 +73,24 @@ def __getitem_hook__(self, key):
         # objects with varying number of attributes
-        return (tuple(self), dict(self.__dict__))
+        kwargs = dict(self.__dict__)
+        kwargs['getters'] = list(kwargs.pop('_getters', []))
```
why  not just `kwargs['getters'] =  list(self.getters)` ?

mloubout commented on devito/types/utils.py:
```diff
@@ -24,6 +24,8 @@ class Stagger(Tag):
 
 class DimensionTuple(EnrichedTuple):
 
+    __rkwargs__ = EnrichedTuple.__rkwargs__ + ('left', 'right')
```
Can add a comment why this needed?

FabioLuporini replied:
if really needed, this would be a hack, which we cannot merge

FabioLuporini commented:
could you also rebase the commit history as a single commit?

let me know if you need help with rebasing

enwask commented:
> if really needed, it would be a hack, which we cannot merge

@FabioLuporini I mean, is it? I would argue rather what's a hack is overriding instance `__dict__` by `EnrichedTuple.__new__`, and adding arbitrary attributes when constructing it and its children (essentially creating anonymous classes). I saw that `left` and `right` were the two attributes used consistently with `DimensionTuple` (child class) so I thought those attributes should explicitly be a part of the class.

I think ideally the hierarchy would be refactored such that those instance hacks aren't used anywhere and the attributes are explicit; that's why I at least thought the change should be made for `DimensionTuple`

FabioLuporini commented:
> > if really needed, it would be a hack, which we cannot merge
> 
> @FabioLuporini I mean, is it? Arguably what's more of a hack is [overriding instance `__dict__` by `EnrichedTuple.__new__`](https://github.com/devitocodes/devito/blob/772f1300b37d345d533e56216414ea1d90795aca/devito/tools/data_structures.py#L48), and adding arbitrary attributes when constructing it and its children (essentially creating anonymous classes).

I think it's perfectly legal to alter `__dict__` within `__new__` for classes that have a dynamic number of kwargs.

That being said, I think I agree with the rest you've said. We should maybe just wrap a dict and pull from it each time we enter `__getattr__`. Sounds neater

enwask commented:
Fair enough. So for now should I leave the explicit `left`/`right` rkwargs for `DimensionTuple` or get rid of them? Probably should have done this before but I can implement an override for `EnrichedTuple._rebuild` to reapply any arbitrary attributes from `__dict__` instead

FabioLuporini reviewed (approved):
for me this is now GTG, thanks. @mloubout ?

mloubout commented:
GTG, need rebase

enwask commented:
Made some changes requested on Slack; refactored `EnrichedTuple._getters` -> `.getters` for cleaner construction, as well as overriding `_rebuild` so arbitrary attributes aren't discarded (previously rebuilding e.g. `DimensionTuple(..., left=whatever)` dropped the additional attributes).

Also added `test_enrichedtuple_rebuild`.
